File: sdk/python/hovel_sdk/payload.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol

PAYLOAD_SCHEMA_V1 = "hovel.payload/v1"
PAYLOAD_RPC_PREFIX = "payload."


@dataclass(frozen=True)
class PayloadTarget:
    os: str
    arch: str
    abi: str = ""
    endianness: str = ""
    minimum_os: str = ""

    def to_rpc(self) -> dict[str, Any]:
        return {
            "os": self.os,
            "arch": self.arch,
            **{
                key: value
                for key, value in (("abi", self.abi), ("endianness", self.endianness), ("minimumOs", self.minimum_os))
                if value
            },
        }


@dataclass(frozen=True)
class PayloadLoadContract:
    execution_model: str
    entry_contract: str = ""
    relocation: str = ""
    dependencies: tuple[str, ...] = ()

    def to_rpc(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "executionModel": self.execution_model,
            **{
                key: value
                for key, value in (("entryContract", self.entry_contract), ("relocation", self.relocation))
                if value
            },
        }
        if self.dependencies:
            result["dependencies"] = list(self.dependencies)
        return result


@dataclass(frozen=True)
class PayloadVariant:
    id: str
    name: str
    version: str
    kind: str
    format: str
    target: PayloadTarget
    load: PayloadLoadContract
    capabilities: tuple[str, ...] = ()
    extensions: dict[str, Any] = field(default_factory=dict)

    def to_rpc(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "id": self.id,
            "name": self.name,
            "version": self.version,
            "kind": self.kind,
            "format": self.format,
            "target": self.target.to_rpc(),
            "load": self.load.to_rpc(),
        }
        if self.capabilities:
            result["capabilities"] = list(self.capabilities)
        if self.extensions:
            result["extensions"] = dict(self.extensions)
        return result


@dataclass(frozen=True)
class PayloadProviderDescriptor:
    provider_id: str
    version: str
    operations: tuple[str, ...]
    payloads: tuple[PayloadVariant, ...] = ()
    extensions: dict[str, Any] = field(default_factory=dict)
    schema_version: str = PAYLOAD_SCHEMA_V1

    def to_rpc(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "schemaVersion": self.schema_version,
            "providerId": self.provider_id,
            "version": self.version,
            "operations": list(self.operations),
        }
        if self.payloads:
            result["payloads"] = [payload.to_rpc() for payload in self.payloads]
        if self.extensions:
            result["extensions"] = dict(self.extensions)
        return result


@dataclass(frozen=True)
class PayloadContent:
    inline_encoding: str = ""
    inline_data: str = ""
    artifact_id: str = ""
    stream_handle: str = ""

    def to_rpc(self) -> dict[str, Any]:
        choices = [bool(self.inline_data), bool(self.artifact_id), bool(self.stream_handle)]
        if sum(choices) != 1:
            raise ValueError("payload content requires exactly one inline, artifact, or stream source")
        if self.inline_data:
            return {"inline": {"encoding": self.inline_encoding, "data": self.inline_data}}
        if self.artifact_id:
            return {"artifact": {"id": self.artifact_id}}
        return {"stream": {"handle": self.stream_handle}}


@dataclass(frozen=True)
class PayloadArtifactV1:
    name: str
    role: str
    variant: PayloadVariant
    media_type: str
    size: int
    sha256: str
    content: PayloadContent
    extensions: dict[str, Any] = field(default_factory=dict)
    schema_version: str = PAYLOAD_SCHEMA_V1

    def to_rpc(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "schemaVersion": self.schema_version,
            "name": self.name,
            "role": self.role,
            "variant": self.variant.to_rpc(),
            "mediaType": self.media_type,
            "size": self.size,
            "sha256": self.sha256,
            "content": self.content.to_rpc(),
        }
        if self.extensions:
            result["extensions"] = dict(self.extensions)
        return result


class PayloadDescriber(Protocol):
    def describe_payloads(self) -> PayloadProviderDescriptor: ...


class PayloadResolver(Protocol):
    def resolve_payload_v1(self, query: dict[str, Any]) -> PayloadVariant: ...


class PayloadGenerator(Protocol):
    def generate_payload_v1(self, request: dict[str, Any]) -> PayloadArtifactV1: ...


class PayloadArtifactReader(Protocol):
    def read_payload_artifact(self, request: dict[str, Any]) -> dict[str, Any]: ...


class PayloadListenerPreparer(Protocol):
    def prepare_payload_listener(self, request: dict[str, Any]) -> dict[str, Any]: ...


class PayloadConnector(Protocol):
    def connect_payload(self, request: dict[str, Any]) -> dict[str, Any]: ...


class PayloadInspector(Protocol):
    def inspect_payload(self, request: dict[str, Any]) -> dict[str, Any]: ...


class PayloadCleanupProvider(Protocol):
    def cleanup_installed_payload(self, request: dict[str, Any]) -> dict[str, Any]: ...

File: sdk/rust/hovel/src/payload.rs
use crate::json::Value;

pub const PAYLOAD_SCHEMA_V1: &str = "hovel.payload/v1";
pub(crate) const PAYLOAD_RPC_DESCRIBE_METHOD: &str = "payload.describe";
pub(crate) const PAYLOAD_RPC_RESOLVE_METHOD: &str = "payload.resolve";
pub(crate) const PAYLOAD_RPC_GENERATE_METHOD: &str = "payload.generate";
pub(crate) const PAYLOAD_RPC_ARTIFACT_READ_METHOD: &str = "payload.artifact.read";
pub(crate) const PAYLOAD_RPC_LISTENER_PREPARE_METHOD: &str = "payload.listener.prepare";
pub(crate) const PAYLOAD_RPC_CONNECT_METHOD: &str = "payload.connect";
pub(crate) const PAYLOAD_RPC_INSPECT_METHOD: &str = "payload.inspect";
pub(crate) const PAYLOAD_RPC_CLEANUP_METHOD: &str = "payload.cleanup";

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PayloadTarget {
    pub os: String,
    pub arch: String,
    pub abi: String,
    pub endianness: String,
    pub minimum_os: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PayloadLoadContract {
    pub execution_model: String,
    pub entry_contract: String,
    pub relocation: String,
    pub dependencies: Vec<String>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PayloadVariant {
    pub id: String,
    pub name: String,
    pub version: String,
    pub kind: String,
    pub format: String,
    pub target: PayloadTarget,
    pub load: PayloadLoadContract,
    pub capabilities: Vec<String>,
    pub extensions: Vec<(String, Value)>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PayloadProviderDescriptor {
    pub schema_version: String,
    pub provider_id: String,
    pub version: String,
    pub operations: Vec<String>,
    pub payloads: Vec<PayloadVariant>,
    pub extensions: Vec<(String, Value)>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum PayloadContent {
    Inline { encoding: String, data: String },
    Artifact { artifact_id: String },
    Stream { handle: String },
}

#[derive(Clone, Debug, PartialEq)]
pub struct PayloadArtifactV1 {
    pub schema_version: String,
    pub name: String,
    pub role: String,
    pub variant: PayloadVariant,
    pub media_type: String,
    pub size: u64,
    pub sha256: String,
    pub content: PayloadContent,
    pub extensions: Vec<(String, Value)>,
}

impl PayloadTarget {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("os", Value::Str(self.os.clone())),
            ("arch", Value::Str(self.arch.clone())),
            ("abi", Value::Str(self.abi.clone())),
            ("endianness", Value::Str(self.endianness.clone())),
            ("minimumOs", Value::Str(self.minimum_os.clone())),
        ])
    }
}

impl PayloadLoadContract {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("executionModel", Value::Str(self.execution_model.clone())),
            ("entryContract", Value::Str(self.entry_contract.clone())),
            ("relocation", Value::Str(self.relocation.clone())),
            (
                "dependencies",
                Value::Array(self.dependencies.iter().cloned().map(Value::Str).collect()),
            ),
        ])
    }
}

impl PayloadVariant {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("id", Value::Str(self.id.clone())),
            ("name", Value::Str(self.name.clone())),
            ("version", Value::Str(self.version.clone())),
            ("kind", Value::Str(self.kind.clone())),
            ("format", Value::Str(self.format.clone())),
            ("target", self.target.to_value()),
            ("load", self.load.to_value()),
            (
                "capabilities",
                Value::Array(self.capabilities.iter().cloned().map(Value::Str).collect()),
            ),
            ("extensions", Value::Object(self.extensions.clone())),
        ])
    }
}

impl PayloadProviderDescriptor {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("schemaVersion", Value::Str(self.schema_version.clone())),
            ("providerId", Value::Str(self.provider_id.clone())),
            ("version", Value::Str(self.version.clone())),
            (
                "operations",
                Value::Array(self.operations.iter().cloned().map(Value::Str).collect()),
            ),
            (
                "payloads",
                Value::Array(self.payloads.iter().map(PayloadVariant::to_value).collect()),
            ),
            ("extensions", Value::Object(self.extensions.clone())),
        ])
    }
}

impl PayloadContent {
    pub(crate) fn to_value(&self) -> Value {
        match self {
            Self::Inline { encoding, data } => Value::object(vec![(
                "inline",
                Value::object(vec![
                    ("encoding", Value::Str(encoding.clone())),
                    ("data", Value::Str(data.clone())),
                ]),
            )]),
            Self::Artifact { artifact_id } => Value::object(vec![(
                "artifact",
                Value::object(vec![("id", Value::Str(artifact_id.clone()))]),
            )]),
            Self::Stream { handle } => Value::object(vec![(
                "stream",
                Value::object(vec![("handle", Value::Str(handle.clone()))]),
            )]),
        }
    }
}

impl PayloadArtifactV1 {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("schemaVersion", Value::Str(self.schema_version.clone())),
            ("name", Value::Str(self.name.clone())),
            ("role", Value::Str(self.role.clone())),
            ("variant", self.variant.to_value()),
            ("mediaType", Value::Str(self.media_type.clone())),
            ("size", Value::from(self.size as i64)),
            ("sha256", Value::Str(self.sha256.clone())),
            ("content", self.content.to_value()),
            ("extensions", Value::Object(self.extensions.clone())),
        ])
    }
}
